### PR TITLE
Go: Stratify `CFG::succ` to avoid recursion

### PR DESCRIPTION
**Pull request by owen-mc**
The first level doesn't deal with defer statements properly. The second level uses the first level to deal with them properly.

### CODE REVIEW

**Comment by owen-mc** [comment]
Happy to take suggestions for a better name to replace `succSimple` for the first layer.

**Comment by owen-mc** [comment]
Performance testing shows no change on the normal set of repos but a bit speedup on the two repos that we noticed the original performance problem on: analysis time went from 1304s to 298s on one and 950s to 228s on the other.

File: go/ql/lib/semmle/go/controlflow/ControlFlowGraphImpl.qll [hunk 1]
```diff
@@ -2074,6 +2080,12 @@ module CFG {
     )
   }
 
+  /** Gets a successor of `nd`, that is, a node that is executed after `nd`. */
```
Comment (smowton) [inline go/ql/lib/semmle/go/controlflow/ControlFlowGraphImpl.qll:2083]: Mention what the `Simple` variant omits

File: go/ql/lib/semmle/go/controlflow/ControlFlowGraphImpl.qll [hunk 2]
```diff
@@ -623,13 +623,15 @@ module CFG {
       not cmpl.isNormal()
     }
 
-    predicate succ(ControlFlow::Node pred, ControlFlow::Node succ) {
+    predicate succSimple(ControlFlow::Node pred, ControlFlow::Node succ) {
```
Comment (smowton) [inline go/ql/lib/semmle/go/controlflow/ControlFlowGraphImpl.qll:626]: Can we come up with a name that say what's different about it? `succWithoutUnwindEdges` perhaps?

Comment (owen-mc) [reply go/ql/lib/semmle/go/controlflow/ControlFlowGraphImpl.qll:626]: `succWithoutExecutingDeferredFunctions`?

Comment (smowton) [reply go/ql/lib/semmle/go/controlflow/ControlFlowGraphImpl.qll:626]: Ah yes, a private `succ0` seems fine, with qldoc noting what's missing. Assuming there's no reason someone outside this module would want succ-but-not-unwind-sequences.

Comment (owen-mc) [reply go/ql/lib/semmle/go/controlflow/ControlFlowGraphImpl.qll]: I can't see why they would want that.

File: go/ql/lib/semmle/go/controlflow/ControlFlowGraphImpl.qll [hunk 3]
```diff
@@ -623,13 +623,20 @@
       not cmpl.isNormal()
     }
 
-    predicate succ(ControlFlow::Node pred, ControlFlow::Node succ) {
+    /**
+     * Holds if `succ` is a successor of `pred`, ignoring the execution of any
+     * deferred functions when a function ends.
+     */
+    predicate succ0(ControlFlow::Node pred, ControlFlow::Node succ) {
```
[the remark this answers is not in the document]
Comment (smowton) [reply go/ql/lib/semmle/go/controlflow/ControlFlowGraphImpl.qll:630]: Hah! This is a correct warning. The magic optimisation restricts a predicate to only be computed where some use-site might use the result -- for example, if I have `private predicate evens(int x) { x % 2 = 0 and x >= 0 and x <= 1000000 } and the only user is `... x <= 10 and evens(x)`, magic might introduce the side-condition `x <= 10` into `evens` and so make computing it cheaper. If that is applied to `succ0` it could at least in theory restrict succ0 to (among other things) defer sequences defined in terms of succ0, and thus reintroduce recursion, though the optimiser should try pretty hard not to create recursion via magic.

Long story short, make this `pragma[nomagic] private`.

[the remark this answers is not in the document]
Comment (owen-mc) [reply go/ql/lib/semmle/go/controlflow/ControlFlowGraphImpl.qll:630]: Sure. This alert is for the member predicate of `ControlFlowTree`. Do I need to do it for the overrides as well? And should I also do it for the top-level predicate?

[the remark this answers is not in the document]
Comment (smowton) [reply go/ql/lib/semmle/go/controlflow/ControlFlowGraphImpl.qll:630]: Good question; not sure how the pragma interacts with overrides (and for that matter, not sure about the combination of private and overrides). Suggest posing the Q to the ql channel.

[the remark this answers is not in the document]
Comment (owen-mc) [reply go/ql/lib/semmle/go/controlflow/ControlFlowGraphImpl.qll:631]: I've asked to be sure, but I think it doesn't get inherited. I've added a commit to add the pragma to all the member predicates. (They can't be private, btw, but they're on a private class.) I'm still unsure about the top level predicate, on line 2115. (Also not sure if it should still be cached, since it's now private and only used to build the top level `succ`.)

File: go/ql/lib/semmle/go/controlflow/ControlFlowGraphImpl.qll [hunk 4]
```diff
@@ -1440,13 +1459,14 @@ module CFG {
         // `defer` can be the first `defer` statement executed
         // there is always a predecessor node because the `defer`'s call is always
         // evaluated before the defer statement itself
-        MkDeferNode(defer) = succ(notDeferSucc*(this.getEntry()))
+        MkDeferNode(defer) = succ0(notDefersucc0*(this.getEntry()))
```
Comment (smowton) [inline go/ql/lib/semmle/go/controlflow/ControlFlowGraphImpl.qll:1462]: Nitpick: notDefersucc0 -> notDeferSucc0

**Review by smowton — COMMENTED** [review]
Indeed there's no need for the private top-level predicate to be cached. As part of a recursive definition along with the member succ0s it should also be nomagic because context from `succ` including `nextDefer` etc is harmful to introduce anywhere in the recursion.

File: go/ql/lib/semmle/go/controlflow/ControlFlowGraphImpl.qll (hunk 4 above)
Comment (owen-mc) [reply go/ql/lib/semmle/go/controlflow/ControlFlowGraphImpl.qll]: Oops

**Review by smowton — APPROVED** [review]
DCA to check the magic changes haven't had any untoward side-effects; then LGTM

**Comment by owen-mc** [comment]
DCA shows no change in performance for the normal repos and the expected speed-ups for the two affected repos (maybe even slightly better - because of to `nomagic` possibly?).
